cwms_radar_api/src/main/java/cwms/radar/data/dao/RatingMetadataDao.java

```diff
@@ -0,0 +1,221 @@
+        super(dsl);
+    }
+
+    public RatingMetadataList retrieve(String cursor, int pageSize, String office,
+                                       String specIdMask) {
+
+        Object[] parts = CwmsDTOPaginated.decodeCursor(cursor);
+        RatingMetadataList.KeySet keySet = RatingMetadataList.KeySet.build(parts);
+
+        pageSize = RatingMetadataList.KeySet.parsePageSize(parts, pageSize);
+
+        Object[] seekValues = null;
+        if (keySet != null) {
+            seekValues = keySet.getSeekValues();
+        }
+        List<RatingMetadata> ratingMetadata = getRatingMetadata(office, specIdMask, seekValues,
+                pageSize);
+
+        RatingMetadataList.Builder builder = new RatingMetadataList.Builder(pageSize);
+        builder.withKeySet(keySet);
+        builder.withMetadata(ratingMetadata);
+
+        return builder.build();
+    }
+
+    @NotNull
+    public List<RatingMetadata> getRatingMetadata(String office, String specIdMask,
+                                                  Object[] seekValues, int pageSize) {
+        List<RatingMetadata> retval;
+
+        AV_RATING_SPEC specView = AV_RATING_SPEC.AV_RATING_SPEC;
+        AV_RATING ratView = AV_RATING.AV_RATING;
+
+        Condition condition = specView.LOC_ALIAS_CATEGORY.isNull()
+                .and(ratView.LOC_ALIAS_CATEGORY.isNull());
+
+        if (office != null) {
+            condition = condition.and(specView.OFFICE_ID.eq(office));
+        }
+
+        if (specIdMask != null) {
+            condition = condition.and(JooqDao.caseInsensitiveLikeRegex(
+                    specView.RATING_ID, specIdMask));
+        }
+
+
+        Collection<OrderField<?>> fields = new ArrayList<>();
+        fields.add(specView.OFFICE_ID);
+        fields.add(specView.TEMPLATE_ID);
+        fields.add(specView.RATING_ID);
+        fields.add(ratView.OFFICE_ID);
+        fields.add(ratView.RATING_ID);
+        fields.add(ratView.EFFECTIVE_DATE.desc());
+
+
+        SelectSeekStepN<Record> records = dsl.select(specView.RATING_SPEC_CODE,
+                        specView.RATING_ID,
+                        specView.TEMPLATE_CODE,
+                        specView.TEMPLATE_ID,
+                        specView.OFFICE_ID,
+                        specView.LOCATION_ID,
+                        specView.DESCRIPTION,
+                        specView.VERSION,
+                        specView.DEP_ROUNDING_SPEC,
+                        specView.IND_ROUNDING_SPECS,
+                        specView.DATE_METHODS,
+                        specView.AUTO_MIGRATE_EXT_FLAG,
+                        specView.AUTO_ACTIVATE_FLAG,
+                        specView.AUTO_UPDATE_FLAG,
+                        specView.ACTIVE_FLAG,
+                        specView.SOURCE_AGENCY,
+                        specView.LOC_ALIAS_GROUP,
+                        specView.LOC_ALIAS_CATEGORY,
+                        specView.ALIASED_ITEM,
+                        ratView.OFFICE_ID,
```

Comment:
this this OFFICE_ID ever be different from the spec OFFICE_ID? if not we also need to keep one.

Reply:
I was just about to dig into the office ids and look at them more closely.  Seems like they should be part of the join, right?

Reply:
generally yes, it helps the planner eliminate a lot of possible rows and speeds things up..

Reply:
I got rid of the extra columns and moved the seek columns into the Keyset so they are together with the values.